core/libp2p/multi/multiaddress/multiaddress.cpp

```diff
@@ -0,0 +1,233 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <algorithm>
+#include <numeric>
+#include <stdexcept>
+
+#include "libp2p/multi/multiaddress.hpp"
+extern "C" {
+#include "libp2p/multi/multiaddress/c-utils/protoutils.h"
+}
+
+namespace {
+  // string representations of protocols
+  constexpr std::string_view kIp4 = "/ip4";
+  constexpr std::string_view kIp6 = "/ip6";
+  constexpr std::string_view kIpfs = "/ipfs";
+  constexpr std::string_view kTcp = "/tcp";
+  constexpr std::string_view kUdp = "/udp";
+  constexpr std::string_view kDccp = "/dccp";
+  constexpr std::string_view kSctp = "/sctp";
+  constexpr std::string_view kUdt = "/udt";
+  constexpr std::string_view kUtp = "/utp";
+  constexpr std::string_view kHttp = "/http";
+  constexpr std::string_view kHttps = "/https";
+  constexpr std::string_view kWs = "/ws";
+  constexpr std::string_view kOnion = "/onion";
+  constexpr std::string_view kWebrtc = "/libp2p-webrtc-star";
+
+  /**
+   * Find beginning of at least one of the provided substrings in the string
+   * @param string to search in
+   * @param substrings to be searched for
+   * @return position of one of the substrings inside of the string or npos, if
+   * none of them is found
+   */
+  size_t findOneOfSubstrings(
+      std::string_view string,
+      std::initializer_list<std::string_view> substrings) {
+    auto pos = std::string_view::npos;
+    std::any_of(std::begin(substrings),
+                std::end(substrings),
+                [&pos, &string](auto substring) {
+                  pos = string.find(substring);
+                  return pos != std::string_view::npos;
+                });
+    return pos;
```

Comment:
if it is just find string in a list, why you don't use
```
auto it = std::find(substrings.begin(), substrings.end(), string);
if (it == substrings.end()) {/*notfound*/}
else {
 return std::distance(substrings.begin(), it);
}
```

Reply:
It is about finding finding one of the strings, which are in the list, in the given string, so the method above would not work

Reply:
Though now I think this function is useless, as its purpose was utilized in the earlier version of the code, so I'm removing it